core/src/main/scala/akka/kafka/internal/SubSourceLogic.scala

```diff
@@ -128,7 +136,15 @@ private class SubSourceLogic[K, V, Msg](
                 )
               )
             case Success(offsets) =>
-              seekAndEmitSubSources(formerlyUnknown, offsets)
+              if (fencingToken == offsetLoadingFencingGen.get()) {
```

Comment:
I think the easiest way to resolve this issue is to create a new `AsyncCallback` `onOffsetsFromExternalResponse`, that's invoked with the response from `getOffsetsFromExternal(assigned)`. In that callback, we can filter partitions out of the offset map returned with what is currently assigned. We'll know what's currently assigned because partitionAssignedCB would have been called concurrently while waiting for the external offset `Future`.

This way we wouldn't require the fencing token and atomic reference.